osu.Game/Database/RealmContextFactory.cs

```diff
@@ -162,6 +170,7 @@ public Realm CreateContext()
             finally
             {
                 contextCreationLock.Release();
+                currentThreadCanCreateContexts.Value = false;
```

Comment:
Shouldn't this be only for the `CreateContext()` leader?

From the docs of [`SemaphoreSlim.Release`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim.release?view=net-6.0#System_Threading_SemaphoreSlim_Release)
> SemaphoreFullException
> The SemaphoreSlim has already reached its maximum size.

Which sounds like the following should crash:
```
using (CreateContext())
{
    // semaphore taken (1)
    // currentThreadcanCreateContexts = true

    using (CreateContext())
    {
        // semaphore not taken
        // currentThreadcanCreateContexts = true
    }

    // Semaphore released (0)
    // currentThreadcanCreateContexts = false

    using (CreateConext()) // Won't this fail?
    {
        // semaphore taken (1)
        // currentThreadcanCreateContexts = true
    }

    // Semaphore released (0)
    // currentThreadcanCreateContexts = false
}

// Semaphore released (-1) <--- Crash??
// currentThreadcanCreateContexts = false
```

Reply:
I've fixed this but I don't think it can happen in practice.

The sample code you have won't trigger the issue in the first place. Note that the lock is not kept for the full `using` block, only the `CreateContext` call. See the call stack I included in the issue thread, it's very subtle.